### ISSUE TITLE
ansible role's dependency is called only once although it is called twice with different parameters.

### ISSUE THREAD

**Issue by suzuki-shunsuke** [ansible/ansible#34736]
##### ISSUE TYPE

 - Bug Report

##### COMPONENT NAME

roles

##### ANSIBLE VERSION

```
2.4.2.0
```

##### OS / ENVIRONMENT

Mac OS X

##### SUMMARY

ansible role's dependency is called only once although it is called twice with defferent parameters.

##### STEPS TO REPRODUCE

roles/foo/meta/main.yml

```yaml
---
dependencies:
  - role: hello
    name: "{{key}}"
```

roles/hello/tasks/main.yml

```yaml
---
- debug:
    var: name
```

playbook.yml

```yaml
---
- hosts: localhost
  roles:
    - role: foo
      key: foo
    - role: foo
      key: bar
```

##### EXPECTED RESULTS

When the playbook is run, the role `hello` is called twice because roles `foo` and `hello` are called twice with different parameters.

http://docs.ansible.com/ansible/latest/playbooks_reuse_roles.html#role-duplication-and-execution

```yaml
PLAY [localhost] *****************************************************************************************************************

TASK [Gathering Facts] ***********************************************************************************************************
ok: [localhost]

TASK [hello : debug] *************************************************************************************************************
ok: [localhost] => {
    "name": "foo"
}

TASK [hello : debug] *************************************************************************************************************
ok: [localhost] => {
    "name": "bar"
}

PLAY RECAP ***********************************************************************************************************************
localhost                  : ok=3    changed=0    unreachable=0    failed=0
```

##### ACTUAL RESULTS

The role `hello` is called only once.

```
PLAY [localhost] *****************************************************************************************************************

TASK [Gathering Facts] ***********************************************************************************************************
ok: [localhost]

TASK [hello : debug] *************************************************************************************************************
ok: [localhost] => {
    "name": "foo"
}

PLAY RECAP ***********************************************************************************************************************
localhost                  : ok=2    changed=0    unreachable=0    failed=0
```

**Comment by alikins** [comment]
This sounds like a similar issue to https://github.com/ansible/ansible/issues/29052

**Comment by alikins** [comment]
pinginh @bcoca, @jimi-c

**Comment by suzuki-shunsuke** [comment]
Reproduce this issue at ansible 2.5.2 .

**Comment by JulienPalard** [comment]
I can reproduce this issue with Ansible 2.9.2.

**Comment by JulienPalard** [comment]
It works using `include_role` instead of `roles`.

**Comment by loop-evgeny** [comment]
I can repro on Ansible 2.9.7. Setting `allow_duplicates: true` on the called role ("hello" in this example) is a workaround.

**Comment by loop-evgeny** [comment]
This seems like a pretty serious bug, which can be difficult to track down in a real, complex playbook. Surprising that nothing has been done about it 2 years later.

**Comment by ansibot** [comment]
@suzuki-shunsuke: Greetings! Thanks for taking the time to open this issue. In order for the community to handle your issue effectively, we need a bit more information.

Here are the items we could not find in your description:
- component name

Please set the description of this issue with an appropriate template from:
https://github.com/ansible/ansible/tree/devel/.github/ISSUE_TEMPLATE

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by Tupteq** [comment]
In 2.10.4 error still exists. As @loop-evgeny said, it seems to be serious error that can be unnoticed in long playbooks. This makes whole `dependencies` mechanism unreliable, and requires from user manual dependency handling. `include_role` or `import_role` aren't real solutions because they don't provide a mechanism allowing to run role only once for each set of parameters.

**Comment by sivel** [comment]
Duplicate of:

* https://github.com/ansible/ansible/issues/72650
* https://github.com/ansible/ansible/issues/67718

**Comment by loop-evgeny** [comment]
#72650 is closed while #67718 is a feature request. This is a bug report, so I'm not sure how it can be a duplicate. Both of those talk about variables past one level of depth, while this issue happens even when constant values passed directly to the role differ.

Not that it really matters, though, since if this serious bug hasn't been fixed in this long it's not going to be.
